bmcs/src/main/scala/akka/stream/alpakka/oracle/bmcs/auth/BmcsSigner.scala

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.alpakka.oracle.bmcs.auth
+
+import java.nio.charset.StandardCharsets
+import java.security.Signature
+import java.security.interfaces.RSAPrivateKey
+import java.time.Instant
+import java.util.Base64
+
+import akka.http.scaladsl.model.headers._
+import akka.http.scaladsl.model._
+import akka.stream.Materializer
+import akka.stream.alpakka.oracle.bmcs.BmcsException
+
+import scala.concurrent.Future
+
+object BmcsSigner {
+
+  val algorithm = SignatureAlgorithm(jvmName = "SHA256withRSA", nameForAuthHeader = "rsa-sha256", version = "1")
+
+  val requestTarget = "(request-target)"
+  val xContentSha256 = "x-content-sha256"
+  val genericHeaders: List[String] = List(Date.lowercaseName, Host.lowercaseName, requestTarget)
+  val bodyHeaders: List[String] = List(`Content-Length`.lowercaseName, `Content-Type`.lowercaseName, xContentSha256)
+  val allHeaders: List[String] = genericHeaders ::: bodyHeaders
+  val requiredHeadersMap: Map[HttpMethod, Seq[String]] = Seq(
+    HttpMethods.GET -> genericHeaders,
+    HttpMethods.HEAD -> genericHeaders,
+    HttpMethods.DELETE -> genericHeaders,
+    HttpMethods.PUT -> genericHeaders,
+    HttpMethods.POST -> allHeaders
+  ).toMap
+
+  def signedRequest(request: HttpRequest,
+                    credentials: BmcsCredentials,
+                    algorithm: SignatureAlgorithm = algorithm,
+                    date: Instant = Instant.now())(
+      implicit mat: Materializer
+  ): Future[HttpRequest] = {
+    import mat.executionContext
+
+    //these headers are just calculated for calculating signature.
+    val pseudoHeaders: Seq[HttpHeader] = calculatePseudoHeaders(request)
+
+    //these headers are actually added to the request. also calculates hash of entity body when needed.
+    val headersToAdd: Future[Seq[HttpHeader]] = calculateHeadersToAdd(request, date)
+    headersToAdd.map { toAdd =>
+      val all = request.headers ++ toAdd ++ pseudoHeaders
+      val authHeader = authorizationHeader(all, request.uri, request.method, algorithm, credentials)
+      request.withHeaders(request.headers ++ toAdd :+ authHeader)
+    }
+  }
+
+  private def calculateHeadersToAdd(request: HttpRequest, date: Instant = Instant.now)(
+      implicit mat: Materializer
+  ): Future[Seq[HttpHeader]] = {
+    import mat.executionContext
+    val dateHdr = Date(DateTime(date.toEpochMilli))
+    request.method match {
+      case HttpMethods.GET | HttpMethods.DELETE | HttpMethods.PUT | HttpMethods.HEAD =>
+        Future.successful(Seq(dateHdr)) //for GET DELETE PUT and HEAD only Date header is added.
+      case HttpMethods.POST =>
+        request.entity.dataBytes
+          .runWith(digest())
+          .map(hash => encodeBase64(hash.toArray))
+          .map(bh => RawHeader(xContentSha256, bh))
+          .map(hd => Seq(hd, dateHdr))
```

Comment:
Does this mean there is an upload method that will have its stream materialized twice? That should be clearly documented on the user API if so, usually when you pass Source[ByteString, _] to some API you likely expect it to materialize that stream only once.

Reply:
documented in the documentation of signedRequest method.